### ISSUE TITLE
AMP command raises BadLocalReturn despite setting requiresAnswer = False

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1985]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#1985|
|Type|defect|
|Created|2006-08-03 19:42:33Z|

I expected that if I set requiresAnswer = False on my amp.command, that I would not have to return any thing from the responder. Instead, BadLocalReturn is raised:
```
twisted.protocols.amp.BadLocalReturn: <bound method SimpleSymmetricCommandProtocol.cmdNoAnswerHello of <SimpleSymmetricCommandProtocol None at 0x-48c78d94>> returned None and <class 'twisted.test.test_amp.NoAnswerHello'> could not serialize it Traceback: exceptions.AttributeError: 'NoneType' object has no attribute 'items'
```

Attachments:

* [amp-requiresanswer.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/86c/86ca4b94b6838eba758fcdd9da31a4c5cc384526/1df6b43b54bd4cadabccf85c93816551d38e5ca5.py) (713 bytes) - added by rwall on 2006-08-03 19:43:50Z - A simple example of how I came across this.
* [amp_requiresanswer.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/86c/86ca4b94b6838eba758fcdd9da31a4c5cc384526/61a49ae7ef986f95929ff8cc0101ece9e0e9a131.diff) (1119 bytes) - added by rwall on 2006-08-03 19:46:34Z - A patch to the amp tests. I f I understand correctly, I don't think the NoAnswerHello command ever gets used because it's responder method is never called.

<details><summary>Searchable metadata</summary>

```
trac-id__1985 1985
type__defect defect
reporter__rwall rwall
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1154634153000000 1154634153000000
changetime__1167070943000000 1167070943000000
version__None None
owner__ 
cc__rwall
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @radix|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

There are really three problems here:

  * Responders for Commands defined to not require a response should be required to return None, but they are not
  * Returning None when a dict is expected or vice versa should result in more well-defined behavior (not a `NoneType has no attribute items' or whatever other random implementation-specific exception happens to come up)
  * The tests are buggy, since they do things which would be disallowed by one of the previous points.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

Man chris is so lazy.  I guess I will have to do this.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

The real issue here is that there is some misunderstanding of how ```requiresAnswer``` is supposed to work.  There is some missing test coverage and documentation, which I will add - but the existing behavior is _almost_ correct.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph removed owner<br>@glyph set status to `new`|
|-|-|

The existing behavior was in fact exactly correct.  Responding to exarkun's three points:

 * Responders for Commands defined not to require a response should return a valid response nonetheless, because ```requiresAnswer``` is an optimization hint that the _client_ can specify, on any request whose response it will not process, to optimize network traffic.
 * Returning None when a dict is expected already logs a precise error, ```BadLocalReturn```, and as the original reporter's bug says, the error message begins with "```<your method> returned <your value> and <your command> could not serialize it```".  It then includes an implementation-specific traceback which could probably be elided for a value of None, but might be useful in the case of more obscure return values which are supposed to masquerade as dictionaries but are not functioning properly.  (More information is better than less, and the more relevant information is already appearing at the front of the log message.)
 * The tests were not buggy - they are all testing valid behavior - but there was no test verifying the behavior of a ```requiresAnswer=False``` when an answer was requested nonetheless.  I've added some tests, and I've added some notes in the documentation to clarify.

Ready for review in ```nofix-1985```.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph commented|
|-|-|

Somebody should still review this, but now that the issue has been analyzed more thoroughly, it is clear that it should not block the release.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_8"></a>|@jerub set owner to @glyph|
|-|-|

Please merge this forward, the docstring _AmpParserBase.callRemote has a conflict.

Thanks for updating the tests and the docstrings. I'm happy with signing off on this after the callRemote docstring is updated. Please merge forward, run the unit tests again, and merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph set owner to @jerub|
|-|-|

Done

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_10"></a>|@radix set owner to @glyph|
|-|-|

Ok, apparently glyph didn't notice that Jerub said it was cool to merge this after fixing the docstring, probably because he left in the review keyword. I'll reassign this back to glyph and delete the review keyword to see what happens...

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_11"></a>|@jerub commented|
|-|-|

Looks good, yeah, merge nofix-1985-2 please.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph set status to `closed`|
|-|-|

(In [19086]) Enhance documentation and testing of ```requiresAnswer``` argument.

The ```requiresAnswer``` attribute of ```Command``` was previously documented as if to suggest that it had symmetric repercussions on the client and server, and the error-checking logic associated with it was not tested anywhere.

This change adds tests for detecting errors and improves the documentation somewhat.

Author: glyph

Reviewer: jerub

Fixes [#1985](https://github.com/twisted/twisted/issues/1985)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
